### PR TITLE
`constrained-generators`: Add `flip_` to avoid having to add new native functions

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
# Description

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

**Review by Soupstraw — APPROVED** [review]

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 1]
```diff
@@ -3816,6 +3860,11 @@ instance (BaseUniverse fn, Member (FunFn fn) fn) => Functions (FunFn fn) fn wher
 
   rewriteRules Id (x :> Nil) = Just x
   rewriteRules (Compose f g) (x :> Nil) = Just $ app f (app g x)
+  -- TODO: this is a bit crippled by the fact that we forget any other rewrite
+  -- rules that we had for `f`. That's something we'll have to think about.
+  rewriteRules (Flip f) (a@Lit {} :> b :> Nil) = Just $ app f b a
+  rewriteRules (Flip f) (a :> b@Lit {} :> Nil) = Just $ app f b a
+  rewriteRules Flip {} _ = Nothing
```
Comment (Soupstraw) [inline libs/constrained-generators/src/Constrained/Base.hs:3868]: Does this mean that a double`flip_` might not converge (as in `flip_` composed with itself)?

Comment (MaximilianAlgehed) [reply libs/constrained-generators/src/Constrained/Base.hs:3868]: It won't cause convergence issues, but it won't simplify away everything you might want simplified away.

Comment (Soupstraw) [reply libs/constrained-generators/src/Constrained/Base.hs:3868]: or maybe something like `flip_ (flip_ f a)` could fail?

Comment (MaximilianAlgehed) [reply libs/constrained-generators/src/Constrained/Base.hs:3868]: Not fail, but not simplify as nicely as you might want

**Pull request merged**